pkg/bootflow/types/state.go

```diff
@@ -0,0 +1,125 @@
+package types
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// State describes a virtual state of a machine being booted
+type State struct {
+	// Do not mutate these values from the outside, the mutability
+	// is owned by State itself.
+
+	SystemArtifacts SystemArtifacts
+	TrustChains     TrustChains
+	CurrentFlow     Flow
+	CurrentStepIdx  uint
+	VerifiedData    []VerifiedData
+}
+
+func typeMapKey(i interface{}) reflect.Type {
+	k := reflect.TypeOf(i)
+	for k.Kind() == reflect.Ptr {
+		k = k.Elem()
+	}
+	return k
+}
+
+func NewState() *State {
+	return &State{
+		SystemArtifacts: map[reflect.Type]SystemArtifact{},
+		TrustChains:     map[reflect.Type]TrustChain{},
+	}
+}
+
+func (state *State) SetFlow(flow Flow, stepIdx uint) {
+	state.CurrentFlow = flow
+	state.CurrentStepIdx = stepIdx
+}
+
+func (state *State) TrustChainExec(
+	trustChainSample TrustChain,
+	callback func(trustChain TrustChain) error,
+) error {
+	key := typeMapKey(trustChainSample)
+	trustChain := state.TrustChains[key]
+	if trustChain == nil {
+		return ErrNoTrustChain{TrustChainKey: key}
+	}
+
+	return callback(trustChain)
+}
+
+func (state *State) EnableTrustChain(trustChain TrustChain) {
+	if reflect.TypeOf(trustChain).Kind() != reflect.Ptr {
+		panic(fmt.Sprintf("%T is not a modifiable type", trustChain))
+	}
+
+	k := typeMapKey(trustChain)
+	if _, ok := state.TrustChains[k]; ok {
+		panic(fmt.Sprintf("double-setting of the same trust chain type: %s", k))
+	}
+	state.TrustChains[k] = trustChain
+}
+
+func (state *State) SystemArtifactExec(
+	systemArtifactSample SystemArtifact,
+	callback func(systemArtifact SystemArtifact) error,
+) error {
+	key := typeMapKey(systemArtifactSample)
+	systemArtifact := state.SystemArtifacts[key]
+	if systemArtifact == nil {
+		return ErrNoSystemArtifact{SystemArtifactKey: key}
+	}
+
+	return callback(systemArtifact)
+}
+
+func (state *State) EnableSystemArtifact(systemArtifact SystemArtifact) {
```

Comment:
I would call AddSystemArtifact :)

Reply:
The problem with `Add` is when we "add apple, and then again add apple" we get "two apples". While here it is not the case: here "each unique apple can be presented only once".

Renamed to `Include`. Hoping now is better :)